seccomp/umount.go

```diff
@@ -117,15 +124,81 @@ func (u *umountSyscallInfo) process() (*sysResponse, error) {
 	return u.tracer.createContinueResponse(u.reqId), nil
 }
 
+// processUmmountAllowed purpose is to prevent immutable resources from being
+// unmounted.
+//
+// Method will return 'true' if the umount operation is deemed legit, and will
+// return 'false' otherwise.
+func (u *umountSyscallInfo) processUmountAllowed(
+	mip domain.MountInfoParserIface) (bool, *sysResponse) {
+
+	// There must be mountinfo state present for this target. Otherwise, let
+	// kernel handle the error.
+	info := mip.GetInfo(u.Target)
+	if info == nil {
+		return true, u.tracer.createContinueResponse(u.reqId)
+	}
+
+	// Return 'eperm' if the umount operation is over an immutable mountpoint.
+	if u.cntr.IsImmutableMountID(info.MountID) {
+		return false, u.tracer.createErrorResponse(u.reqId, syscall.EPERM)
+	}
+
+	// At this point we have already concluded that the mountpoint being
+	// unmounted doesn't match any of the immutable resources, so we could
+	// be tempted to allow all umount instructions reaching this stage to
+	// succeed. However, there is one particular case that we must identify
+	// to prevent the exposure of immutable masked resources. This is the case
```

Comment:
"This is the case of an umount instruction launched from a different mount namespace targeting an immutable resource (e.g., unshare -m umount /var/lib/docker), because that mount namespace has a different set of mount IDs. The rest of the code ..."

Reply:
I'll rewrite it to incorporate some of this, but keep in mind that we are always talking about an umount instruction generated form within an unshared() context; all other cases that I can think of involve inner containers, and for those, a chroot/pivot_root instruction is involved, which prevent this corner-case scenario from happening.